### PR TITLE
[IOS-424]: Leave group support

### PR DESCRIPTION
**Pull request by yuliaveres**
So we agreed with Marek to use Leave instead of Exit:
img_0024
img_0025
img_0027

### CODE REVIEW

File: Toshi/Controllers/Chat/ChatInteractor.swift [hunk 1]
```diff
@@ -79,8 +79,15 @@ final class ChatInteractor: NSObject {
 
         messageSender?.sendAttachmentData(datasource, contentType: "image/jpeg", sourceFilename: "image.jpeg", in: outgoingMessage, success: {
```
Comment (designatednerd) [inline Toshi/Controllers/Chat/ChatInteractor.swift]: Good grief, that's obnoxious that this doesn't come back on the main thread by default 😠

File: Toshi/Controllers/Chat/ChatInteractor.swift [hunk 2]
```diff
@@ -360,13 +367,51 @@ final class ChatInteractor: NSObject {
...
-                interactor.send(image: groupAvatar, in: outgoingMessage)
+                interactor.send(image: groupAvatar, in: outgoingMessage, completion: completion)
             } else {
-                interactor.send(outgoingMessage)
+                interactor.send(outgoingMessage, completion: completion)
             }
         }
     }
 
+    static func deleteThread(_ thread: TSThread) {
```
Comment (designatednerd) [inline Toshi/Controllers/Chat/ChatInteractor.swift]: What about making this an extension method on `TSThread`? Might help us get some of the database logic out of the VC

File: Toshi/Controllers/GroupChat/GroupInfo/GroupInfoViewModel.swift [hunk 3]
```diff
@@ -51,11 +59,23 @@ final class GroupInfoViewModel {
...
+        models = [avatarTitleSectionData, settingsSectionData, participantsSectionData, exitGroupSectionData]
+    }
+
+    private func setupParticipantsSection() -> TableSectionData {
         let addParticipantsData = TableCellData(title: Localized("new_group_add_participants_action_title"))
         addParticipantsData.tag = GroupItemType.addParticipant.rawValue
 
         var participantsCellData: [TableCellData] = [addParticipantsData]
-        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { return }
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else { fatalError("Can't access app delegate") }
```
Comment (designatednerd) [inline Toshi/Controllers/GroupChat/GroupInfo/GroupInfoViewModel.swift]: should this crash the app at runtime even in production? And if it should, should we fire a log off to Crashlytics before we do?

File: Toshi/Resources/Localizable.strings [hunk 4]
```diff
@@ -266,6 +266,11 @@
 "new_group_public_settings_title" = "Public";
 "new_group_notifications_settings_title" = "Notifications";
 
+"group_info_exit_group" = "Exit group";
+"group_info_exit_confirmation_message" = "Do you want to exit group?";
```
Comment (designatednerd) [inline Toshi/Resources/Localizable.strings]: Maybe "Do you want to exit **this** group?" or "Do you want to exit [group name]?" - this seems worded a little oddly.

File: Toshi/Resources/Localizable.strings [hunk 5]
```diff
@@ -266,6 +266,11 @@
...
+"group_info_exit_action_title" = "Exit";
+"group_info_leave_group_failure_message" = "Failed to leave group";
```
Comment (designatednerd) [inline Toshi/Resources/Localizable.strings]: Is there anything the user can do once they've failed to leave the group?

File: Toshi/Controllers/GroupChat/GroupViewModelProtocol.swift [hunk 6]
```diff
@@ -28,6 +29,8 @@ struct GroupInfo {
 protocol GroupViewModelCompleteActionDelegate: class {
 
     func groupViewModelDidFinishCreateOrUpdate()
+    func groupViewModelDidStartCreateOrUpdate()
+    func groupViewModelaDidRequireReload(_ viewModel: GroupViewModelProtocol)
```
Comment (designatednerd) [inline Toshi/Controllers/GroupChat/GroupViewModelProtocol.swift]: looks like a spare `a` snuck in before `Did`

File: Toshi/Controllers/GroupChat/GroupInfo/GroupInfoViewModel.swift (hunk 3 above)
Comment (yuliaveres) [reply Toshi/Controllers/GroupChat/GroupInfo/GroupInfoViewModel.swift]: lets log to Crashlytics ! (it should)

File: Toshi/Controllers/Chat/ChatInteractor.swift (hunk 2 above)
Comment (yuliaveres) [reply Toshi/Controllers/Chat/ChatInteractor.swift]: 🤔

File: Toshi/Resources/Localizable.strings (hunk 4 above)
Comment (yuliaveres) [reply Toshi/Resources/Localizable.strings]: this group ..  i will check with Marek right now

File: Toshi/Controllers/Chat/ChatInteractor.swift (hunk 2 above)
Comment (yuliaveres) [reply Toshi/Controllers/Chat/ChatInteractor.swift]: that is not controller though. i try to keep all messages / group related logic just in the interactor

Comment (designatednerd) [reply Toshi/Controllers/Chat/ChatInteractor.swift]: Oh yeah derp, my bad.

File: Toshi/Resources/Localizable.strings (hunk 5 above)
Comment (yuliaveres) [reply Toshi/Resources/Localizable.strings]: i do not think so. I would guess dismiss the alert and go back or press exit once more ..

Comment (designatednerd) [reply Toshi/Resources/Localizable.strings]: maybe add something along the lines of "Please try again" then

Comment (yuliaveres) [reply Toshi/Resources/Localizable.strings]: ok. lets do this

**Pull request merged**